nixos/modules/services/networking/nbd.nix

```diff
@@ -0,0 +1,152 @@
+      else { }));
+    }
+    // (mapAttrs
+      (_: { path, allowAddresses, extraOptions }: {
+        exportname = path;
+      }
+      // extraOptions
+      // (if (allowAddresses != null)
+      then {
+        authfile = pkgs.writeText "authfile" ''
+          ${concatStringsSep "\n" allowAddresses}
+        '';
+      } else { }))
+      cfg.server.exports)
+    );
+  allowedDevices =
+    concatLists
+      (mapAttrsToList
+        (_: { path, ... }: if hasPrefix "/dev/" path then [ "${path} rw" ] else [ ])
+        cfg.server.exports);
+in
+{
+  options = {
+    services.nbd = {
+      server = {
+        enable = mkEnableOption "the Network Block Device (nbd) server";
+
+        listenPort = mkOption {
+          type = types.port;
+          default = 10809;
+          description = "Port to listen on. The port is NOT automatically opened in the firewall.";
+        };
+
+        extraOptions = mkOption {
+          type = iniFields;
+          default = {
+            allowlist = false;
+          };
+          description = ''
+            Extra options for the server. See
+            <citerefentry><refentrytitle>nbd-server</refentrytitle>
+            <manvolnum>5</manvolnum></citerefentry>.
+          '';
+        };
+
+        exports = mkOption {
+          description = "Files or block devices to make available over the network.";
+          default = { };
+          type = with types; attrsOf
+            (submodule {
+              options = {
+                path = mkOption {
+                  type = types.str;
+                  description = "File or block device to export.";
+                  example = "/dev/sdb1";
+                };
+
+                allowAddresses = mkOption {
+                  type = with types; nullOr (listOf str);
+                  default = null;
+                  example = [ "10.10.0.0/24" "127.0.0.1" ];
+                  description = "IPs and subnets that are authorized to connect for this device. If not specified, the server will allow all connections.";
+                };
+
+                extraOptions = mkOption {
+                  type = iniFields;
+                  default = {
+                    flush = true;
+                    fua = true;
+                  };
+                  description = ''
+                    Extra options for this export. See
+                    <citerefentry><refentrytitle>nbd-server</refentrytitle>
+                    <manvolnum>5</manvolnum></citerefentry>.
+                  '';
+                };
+              };
+            });
+        };
+
+        listenAddress = mkOption {
+          type = with types; nullOr str;
+          description = "Address to listen on. If not specified, the server will listen on all interfaces.";
+          default = null;
+          example = "10.10.0.1";
+        };
+      };
+    };
+  };
+
+  config = mkIf cfg.server.enable {
+    boot.kernelModules = [ "nbd" ];
```

Comment:
is this necessary on the server? thought that was a userspace-only service.

Reply:
The server needs the `nbd` module.  If it wasn't loaded, the server would just load it itself... until I set `ProtectKernelModules=true` and now it's not allowed to.  So, we need to do it manually.

Reply:
ah, wasn't aware of that. :)